### PR TITLE
Support print `ChainExpression`

### LINKED ISSUE #10244: Refactor: Print `ChainExpression` directly
Currently, we are [transforming `ChainElement`s to `OptionalCallExpression` and `OptionalMemberExpression`](https://github.com/prettier/prettier/blob/8c5523882befca27da702bd5fd97e56bcca54b2a/src/language-js/parse-postprocess.js#L195-L212).

It's supposed to be a temporary workaround for `espree` and `meriyah` parser, now `typescript` parser use this AST type too, [it's a standard `estree` AST](https://github.com/estree/estree/blob/master/es2020.md#chainexpression), we should support print it directly.

This refactor may need a big change in `CallExpression` `MemberExpression` related logic part.

### PR DESCRIPTION
**Pull request by fisker**
## Description

Fixes #10244

## Checklist

- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [ ] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/language-js/needs-parens.js [hunk 1]
```diff
@@ -687,6 +687,37 @@ function needsParens(path, options) {
...
+              !node.optional)
+        ) ||
+          path.match(
+            undefined,
+            undefined,
+            (node, name) =>
+              name === "expression" && node.type === "TSNonNullExpression",
+            (node, name) =>
+              name === "object" && node.type === "MemberExpression"
+          ))
```
Comment (sosukesuzuki) [inline src/language-js/needs-parens.js:715]: It seems unreadable to me... Can you add comment or extract these conditions to variables?

Comment (fisker) [reply src/language-js/needs-parens.js:715]: What kind of comment do you expect? Similar code exists

https://github.com/prettier/prettier/blob/67e121a20b9d5d983970bdc2dcba5de2bd0f7beb/src/language-js/print/function-parameters.js#L246
https://github.com/prettier/prettier/blob/67e121a20b9d5d983970bdc2dcba5de2bd0f7beb/src/language-js/print/object.js#L200

Comment (sosukesuzuki) [reply src/language-js/needs-parens.js:725]: I expect comments that describes what concrete syntax each matcher represents.

Comment (fisker) [reply src/language-js/needs-parens.js]: [`e849c94`](https://github.com/prettier/prettier/pull/13735/commits/e849c94bbf45d4ceb502402fadf2ea425b5e4dc1)

**Review by sosukesuzuki — APPROVED** [review]

**Pull request merged**